Add reusable blocks data layer

noisysocks opened the pull request:
## Description

Adds the actions, selectors and reducers necessary for supporting reusable blocks. There are 5 actions:

- `FETCH_REUSABLE_BLOCKS`—Loads reusable blocks from the API and inserts them into the store.
- `UPDATE_REUSABLE_BLOCK`—Inserts and updates a reusable block that is in the store.
- `SAVE_REUSABLE_BLOCK`—Persists a reusable block that's in the store to the API.
- `MAKE_BLOCK_STATIC`—Transforms a reusable block on the page into a regular block.
- `MAKE_BLOCK_REUSABLE`—Transforms a regular block on the page into a reusable block.

https://github.com/WordPress/gutenberg/issues/1516 describes the feature and https://github.com/WordPress/gutenberg/issues/2081 (https://github.com/WordPress/gutenberg/issues/2081#issuecomment-325231157, in particular) describes the technical approach I'm taking.

This PR is a subset of https://github.com/WordPress/gutenberg/pull/2659, which was too large. Check it out though if you'd like to play with the complete feature!

## How Has This Been Tested?

Unit tests are included for the new actions, selectors and reducers. 

I changed the inserter to ignore blocks that are marked with `isPrivate: true`, so it's worth testing the inserter to make sure that there are no regressions.

---
cc. @mtias @aduth @youknowriad 

❤️

noisysocks committed:
Add reusable blocks data layer

Adds the actions, selectors and reducers necessary for supporting reusable
blocks. There are 5 actions:

- FETCH_REUSABLE_BLOCKS: Loads reusable blocks from the API and inserts
  them into the store.
- UPDATE_REUSABLE_BLOCK: Inserts of updates a reusable block that is in
  the store.
- SAVE_REUSABLE_BLOCK: Persists a reusable block that's in the store to
  the API.
- MAKE_BLOCK_STATIC: Transforms a reusable block on the page into a
  regular block.
- MAKE_BLOCK_REUSABLE: Transforms a regular block on the page into a
  reusable block.

youknowriad commented on editor/actions.js:
```diff
@@ -386,3 +386,74 @@ export const createSuccessNotice = partial( createNotice, 'success' );
+ * @param {string} id If given, only a single reusable block with this ID will be fetched
+ * @return {Object}   Action object
+ */
+export function fetchReusableBlocks( id = null ) {
```
Minor: do we really need the default value?

noisysocks replied:
I suppose not! Removed in 7762cb73a84b08f60938bc004cfd1fa20bccf36a.

youknowriad commented on editor/reducer.js:
```diff
@@ -546,6 +546,65 @@ export function notices( state = {}, action ) {
+				...newState,
+				[ reusableBlock.id ]: {
+					...reusableBlock,
+					isSaving: false,
```
Why are we merging the `isSaving` flag with thee reusable block object, my first reaction here is: "Show this be in a separate reducer (or subReducer)?

noisysocks replied:
No reason other than that it's convenient to pass the block's data and metadata (`isSaving`) to UI components all in one go.

I'd be happy to move this to a different reducer. Perhaps the shape of the redux tree could be something like:
```js
reusableBlocks: {
    data: {
        'f5ca58b0-aad0-4777-bb10-f3d0fa08a163': /* the block */,
    },
    isSaving: {
        'f5ca58b0-aad0-4777-bb10-f3d0fa08a163': false,
    },
}
```

youknowriad replied:
> Perhaps the shape of the redux tree could be something like

I like this proposal. Request flags could be generic, it makes sense to me to separate them from the data.

noisysocks replied:
👌 done in fb9d4c70a898908eddcf9b71028d16262d171791.

youknowriad reviewed (commented):
Left some small comments, but this is in a good shape. Nice work

noisysocks committed:
Remove default param in lieu of documenting the param as optional

noisysocks committed:
Seperate metadata (isSaving) from reusable block data in the Redux tree

Re-shape the Redux state tree so that metadata like `isSaving` is kept
seperately from each reusable block's actual data.

youknowriad commented on editor/reducer.js:
```diff
@@ -562,64 +562,58 @@ export function notices( state = {}, action ) {
+				return reduce( action.reusableBlocks, ( newState, reusableBlock ) => ( {
+					...newState,
+					[ reusableBlock.id ]: {
+						...reusableBlock,
```
Is there a reason we're still cloning the object?

youknowriad commented on editor/reducer.js:
```diff
@@ -562,64 +562,58 @@ export function notices( state = {}, action ) {
-			const { id } = action;
+	isSaving( state = {}, action ) {
+		switch ( action.type ) {
+			case 'SAVE_REUSABLE_BLOCK': {
```
I guess the scoping (`{`) is not useful anymore

youknowriad reviewed (approved):
Nice work 👍

mtias commented on blocks/api/categories.js:
```diff
@@ -17,6 +17,7 @@ const categories = [
 	{ slug: 'layout', title: __( 'Layout Blocks' ) },
 	{ slug: 'widgets', title: __( 'Widgets' ) },
 	{ slug: 'embed', title: __( 'Embed' ) },
+	{ slug: 'reusable-blocks', title: __( 'My Reusable Blocks' ) },
```
These should probably be allocated in another "tab" (like recent, blocks, embeds) rather than a category. The names should probably be "Saved". cc @jasmussen

youknowriad committed:
Reusable Block Reducer: small tweaks

aduth commented on blocks/api/factory.js:
```diff
@@ -116,3 +121,19 @@ export function switchToBlockType( block, name ) {
+ */
+export function createReusableBlock( type, attributes ) {
+	return {
+		id: uuid(),
```
The apparent overlap yet disconnect between a reusable block's `id` property and a standard block's `uid` property is unfortunate. I can see why we might want to call this one `id`, since it aligns well with how it is modeled on the server (the post's `id`), but it makes me wonder whether we should then align the standard block property to `id` as well. I'm not strongly attached to `uid` as a property name.

Do you have any thoughts?

noisysocks replied:
I'd be into aligning them so long as we're explicit when writing code that interacts with both types of ID, e.g:

```js
const blockId = block.id;
const reusableBlockId = block.attributes.ref;
```

This would be a breaking change though, since there might be third party code that references `block.uid`.

aduth commented on blocks/api/factory.js:
```diff
@@ -116,3 +121,19 @@ export function switchToBlockType( block, name ) {
+export function createReusableBlock( type, attributes ) {
+	return {
+		id: uuid(),
+		name: __( 'Untitled block' ),
```
Would there ever be a case where we want to create a reusable block and assign its name immediately? A bit awkward to achieve now, since this is not accepted as an argument of the function.

noisysocks replied:
No such case was in our design so I figured _you aren't gonna need it_. We could add an optional argument in the future if it comes up.

mtias replied:
I think we should support this and have UI that focus on the name input before saving. cc @jasmussen
